### PR TITLE
Properly handle null as system default proxy selector (#4232)

### LINKED ISSUE #4232: NullPointerException if default ProxySelector is null
Here the test:

```java
private final MockWebServer server = new MockWebServer();
@Before public void setUp() throws Exception { server.start(); }
@After public void tearDown() throws Exception { server.shutdown(); }
@Test public void nullDefaultProxySelector() throws Exception {
  server.enqueue(new MockResponse().setBody("abc"));

  ProxySelector.setDefault(null);

  OkHttpClient client = defaultClient().newBuilder()
      .build();

  Request request = new Request.Builder().url(server.url("/")).build();
  Response response = client.newCall(request).execute();
  assertEquals("abc", response.body().string());
}
```

And the result:

```java
Aug 25, 2018 4:36:39 PM okhttp3.mockwebserver.MockWebServer$2 execute
INFORMATION: MockWebServer[57774] starting to accept connections
Aug 25, 2018 4:36:39 PM okhttp3.mockwebserver.MockWebServer$2 acceptConnections
INFORMATION: MockWebServer[57774] done accepting connections: socket closed
Aug 25, 2018 4:36:39 PM okhttp3.SocksProxy$1 execute
INFORMATION: SocksProxy 57775 done accepting connections: socket closed

java.lang.NullPointerException: proxySelector == null

	at okhttp3.Address.<init>(Address.java:79)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.createAddress(RetryAndFollowUpInterceptor.java:212)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:112)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at okhttp3.SocksProxyTest.nullDefaultProxySelector(SocksProxyTest.java:101)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

To quote https://docs.oracle.com/javase/8/docs/api/java/net/ProxySelector.html#setDefault-java.net.ProxySelector-: `ps - The HTTP proxy selector, or null to unset the proxy selector.`, so `null` is an acceptable value that OkHttp should be able to cope with.

### PR DESCRIPTION
**Pull request by Vampire**
Fixes #4232

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/OkHttpClient.java [hunk 1]
```diff
@@ -460,6 +463,21 @@ public Builder() {
       connectionSpecs = DEFAULT_CONNECTION_SPECS;
       eventListenerFactory = EventListener.factory(EventListener.NONE);
       proxySelector = ProxySelector.getDefault();
+      if (proxySelector == null) {
+        proxySelector = new ProxySelector() {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/OkHttpClient.java]: I'd suggest moving this to a named class under the internal package.  No reason to have this logic inside OkHttpClient, it's big enough already.

Comment (Vampire) [reply okhttp/src/main/java/okhttp3/OkHttpClient.java]: That actually was my first thought. :-)
I was just not sure where exactly and how to name it to fit your conventions, so I made it like this waiting for complaints. :-D
`internal` directly? Or `internal/connection`? Or `internal/proxy`?
What would be a good class name that fits your naming policies?
`NoProxySelector`?

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/OkHttpClient.java]: I liked NullProxySelector that @swankjesse suggested. Either internal/http or internal/proxy would be fine for me.  No strong opinion from me.

**Comment by yschimke** [comment]
Thanks for the report and the fix! I think you have already done it, but if not, please submit the CLA here http://square.github.io/okhttp/#contributing

File: okhttp/src/main/java/okhttp3/OkHttpClient.java (hunk 1 above)
Comment (Vampire) [reply okhttp/src/main/java/okhttp3/OkHttpClient.java]: Ok, then I think `internal/proxy` is better fit.
Well, I don't like `NullProxySelector` as it does not select a `null` proxy, it selects the `NO_PROXY` proxy that stands for a direct connection.
That it is used if the system default proxy selector is `null` is not really significant imho.
But if you prefer that name, I'll change it to that, it's your project. :-)

**Comment by Vampire** [comment]
I cannot remember having done it.
Either way, I did now.

File: okhttp/src/main/java/okhttp3/internal/proxy/NullProxySelector.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package okhttp3.internal.proxy;
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/proxy/NullProxySelector.java]: Please add the copyright notice.

Comment (Vampire) [reply okhttp/src/main/java/okhttp3/internal/proxy/NullProxySelector.java]: Whoops, my bad, fixed

**Pull request merged**